File: C++/Decode Ways.cpp
#include <iostream>
#include <string>

using namespace std;

int numDecodings(string s) {
    int len=s.size();
    if(len==0) return 0;
    int *dp=new int[len+1];
    fill(dp,dp+len+1,0);
    dp[0]=1;
    if(s[0]!='0') dp[1]=1;
    for(int i=2;i<=len;i++){
        int cnt=(s[i-2]-'0')*10+s[i-1]-'0';
        if(s[i-1]=='0'&&cnt<=26&&cnt>0){
            dp[i]=dp[i-2];
        }
        else if(s[i-1]!='0'&&s[i-2]!='0'&&cnt<=26&&cnt>0){
            dp[i]=dp[i-2]+dp[i-1];
        }
        else if(s[i-1]!='0'){
            dp[i]=dp[i-1];
        }
    }
    return dp[len];
}

File: Python/Decode-Ways.py
class Solution:
    # @param s, a string
    # @return an integer
    def numDecodings(self, s):
        length = len(s)
        if length == 0:
            return 0
        dp = [0] * (length + 1)
        dp[0] = 1
        if s[0] != '0':
            dp[1] = 1
        for i in range(2, length+1):
            cnt = int(s[i-2]) * 10 + int(s[i-1])
            if s[i-1] == '0' and cnt <= 26 and cnt > 0:
                dp[i] = dp[i-2]
            elif s[i-1] != '0' and s[i-2] != '0' and cnt <= 26 and cnt > 0:
                dp[i] = dp[i-2] + dp[i-1]
            elif s[i-1] != '0':
                dp[i] = dp[i-1]
        return dp[length]